optpresso/server/static/js/app.js

```diff
@@ -0,0 +1,394 @@
+  xmlHttp.send(formData);
+}
+
+function httpGetAsync(url, callback)
+{
+  var xmlHttp = new XMLHttpRequest();
+  xmlHttp.onreadystatechange = function() { 
+    if (xmlHttp.readyState == 4 && xmlHttp.status % 200 < 100)
+      callback(xmlHttp.responseText);
+  }
+  xmlHttp.open("GET", url, true); // true for asynchronous
+  xmlHttp.send();
+}
+
+class Video extends React.Component {
+  componentDidMount() {
+    var videoWidth, videoHeight;
+
+    // whether streaming video from the camera.
+    var streaming = false;
+
+    var video = document.getElementById('video');
+    var canvasOutput = document.getElementById('canvasOutput');
+    var canvasOutputCtx = canvasOutput.getContext('2d');
+    var stream = null;
+
+    function startCamera() {
+      if (streaming) return;
+      navigator.mediaDevices.getUserMedia({video: true, audio: false})
+        .then(function(s) {
+          stream = s;
+          video.srcObject = s;
+          video.play();
+      })
+        .catch(function(err) {
+          console.log("An error occured! " + err);
+      });
+
+      video.addEventListener("canplay", function(ev){
+        if (!streaming) {
+          videoWidth = video.videoWidth;
+          videoHeight = video.videoHeight;
+          video.setAttribute("width", videoWidth);
+          video.setAttribute("height", videoHeight);
+          canvasOutput.width = videoWidth;
+          canvasOutput.height = videoHeight;
+          streaming = true;
+        }
+        startVideoProcessing();
+      }, false);
+    }
+
+    var canvasInput = null;
+    var canvasInputCtx = null;
+
+    var canvasBuffer = null;
+    var canvasBufferCtx = null;
+
+    function startVideoProcessing() {
+      if (!streaming) { console.warn("Please startup your webcam"); return; }
+      canvasInput = document.createElement('canvas');
+      canvasInput.width = videoWidth;
+      canvasInput.height = videoHeight;
+      canvasInputCtx = canvasInput.getContext('2d');
+      
+      canvasBuffer = document.createElement('canvas');
+      canvasBuffer.width = videoWidth;
+      canvasBuffer.height = videoHeight;
+      canvasBufferCtx = canvasBuffer.getContext('2d');
+     
+      
+      requestAnimationFrame(processVideo);
+    }
+
+    function processVideo() {
+      canvasInputCtx.drawImage(video, 0, 0, videoWidth, videoHeight);
+      var imageData = canvasInputCtx.getImageData(0, 0, videoWidth, videoHeight);
+      canvasOutputCtx.drawImage(canvasInput, 0, 0, videoWidth, videoHeight);
+      requestAnimationFrame(processVideo);
+    }
+
+    startCamera()
+  }
+
+  render() {
+    return (
+      <div>
+        <div class="container">
+          <canvas class="center-block" id="canvasOutput"></canvas>
+        </div>
+        <div class="invisible">
+          <video id="video" class="hidden">Your browser does not support the video tag.</video>
+        </div>
+      </div>
+    )
+  }
+}
+
+
+class Nav extends React.Component {
```

Comment:
Ahem, there is something called a "file" that could be quite useful here.. 👯‍♂️

Reply:
Not sure what you mean by that?

Reply:
Haha just that you could put components into different files.. so it wasn't just all in one single file

Reply:
Ah yes, mostly this is a first pass hack at making this into 'something' and so ONE FILE to rule them all